File: src/ai/role-settings.ts
/**
 * 実行時のモデル設定解決(チャット/執筆系/判断系/バックグラウンド)。
 * main.ts に私有されていた解決ロジックを、設定モーダルのプレビューなど
 * 他モジュールからも同じ結果を計算できるよう共有化したもの。
 * モジュール変数には依存せず、ProviderConfig を引数で受け取る。
 */

import {
  resolveChatSettings,
  resolveBackgroundSettings,
  resolveWritingSettings,
  resolveJudgmentSettings,
  type AiSettings,
  type AnthropicThinkingEffort,
  type RoleParamOverrides,
} from "../settings.ts";
import {
  getProviderEntry,
  getProviderModelDefaults,
  type ModelRoleProfile,
  type ProviderConfig,
  type ProviderModelDefaults,
} from "../providers/config.ts";
import { getEffectiveCapability, isDeepSeekV4Model } from "./capability.ts";
import { getCopilotModelCacheEntry } from "../providers/copilot-auth.ts";

export const NON_THINKING_WRITING_MAX_TOKENS = 8192;

export function enforceRequiredThinking(settings: AiSettings): AiSettings {
  return isDeepSeekV4Model(settings.model)
    ? { ...settings, deepseekThinkingEnabled: true }
    : settings;
}

export function isThinkingEnabledForRun(settings: AiSettings): boolean {
  if (isDeepSeekV4Model(settings.model)) return true;
  const capability = settings.reasoningCapability;
  if (capability?.kind === "deepseek" || (!capability && settings.provider === "deepseek")) {
    return settings.deepseekThinkingEnabled !== false;
  }
  if (capability?.kind === "anthropic-adaptive") return true;
  if (capability?.kind === "anthropic-budget" || (!capability && settings.provider === "anthropic")) {
    return settings.anthropicThinkingEnabled === true;
  }
  if (capability?.kind === "openai" || (!capability && (settings.provider === "openai" || settings.provider === "codex"))) {
    return Boolean(settings.openaiReasoningEffort && settings.openaiReasoningEffort !== "none");
  }
  if (capability?.kind === "google") return true;
  // OpenCode GoのV4系は明示的なThinking OFFを現在提供しておらず、既定挙動を維持する。
  if (settings.provider === "opencode") return true;
  return false;
}

export function applyRuntimeModelDefaults(
  settings: AiSettings,
  defaults: ProviderModelDefaults | undefined,
  applyTokenDefaults: boolean,
): AiSettings {
  if (!defaults) return settings;

  const capability = getEffectiveCapability(
    settings.provider,
    settings.model,
    defaults,
    getCopilotModelCacheEntry,
  );
  const next: AiSettings = {
    ...settings,
    temperature: defaults.temperature ?? settings.temperature,
    maxTokens: applyTokenDefaults ? defaults.maxTokens ?? settings.maxTokens : settings.maxTokens,
    maxContextTokens: applyTokenDefaults ? defaults.maxContextTokens ?? settings.maxContextTokens : settings.maxContextTokens,
    topP: defaults.topP,
    topK: defaults.topK,
    frequencyPenalty: defaults.frequencyPenalty,
    presencePenalty: defaults.presencePenalty,
    reasoningCapability: capability,
  };
  if (capability?.kind === "openai") {
    next.openaiReasoningEffort = defaults.openaiReasoningEffort ?? settings.openaiReasoningEffort;
  } else if (capability?.kind === "deepseek") {
    next.deepseekReasoningEffort = defaults.deepseekReasoningEffort ?? settings.deepseekReasoningEffort;
  } else if (capability?.kind === "anthropic-adaptive") {
    next.anthropicThinkingEnabled = true;
    next.anthropicThinkingBudget = undefined;
    const defaultEffort = capability.defaultEffort;
    const validDefaultEffort =
      defaultEffort === "low" || defaultEffort === "medium" || defaultEffort === "high" ||
      defaultEffort === "xhigh" || defaultEffort === "max"
        ? defaultEffort as AnthropicThinkingEffort
        : undefined;
    next.anthropicThinkingEffort = defaults.anthropicThinkingEffort ?? settings.anthropicThinkingEffort ?? validDefaultEffort;
  } else if (capability?.kind === "anthropic-budget") {
    next.anthropicThinkingEnabled = defaults.anthropicThinkingEnabled ?? settings.anthropicThinkingEnabled;
    next.anthropicThinkingBudget = defaults.anthropicThinkingBudget ?? settings.anthropicThinkingBudget;
    next.anthropicThinkingEffort = undefined;
  } else if (capability?.kind === "google") {
    next.googleThinkingLevel = defaults.googleThinkingLevel ?? settings.googleThinkingLevel;
  }
  return next;
}

/**
 * 役割プロファイル(providers.json の writing/judgment)またはユーザーの役割別
 * オーバーライドを1レイヤー分だけ設定へ重ねる内部ヘルパー。
 * サンプリング系パラメータと promptScaffold は無条件で上書きする。
 * reasoning/thinking 系は applyRuntimeModelDefaults と同じ方式で、現在のプロバイダに
 * 一致するフィールドだけを適用する(不一致のフィールドは無視するだけで、
 * 対象プロバイダのフィールドを undefined 化はしない)。
 */
function applyRoleProfileLayer(settings: AiSettings, layer: ModelRoleProfile | undefined): AiSettings {
  if (!layer) return settings;
  const next: AiSettings = { ...settings };

  if (layer.temperature !== undefined) next.temperature = layer.temperature;
  if (layer.topP !== undefined) next.topP = layer.topP;
  if (layer.topK !== undefined) next.topK = layer.topK;
  if (layer.frequencyPenalty !== undefined) next.frequencyPenalty = layer.frequencyPenalty;
  if (layer.presencePenalty !== undefined) next.presencePenalty = layer.presencePenalty;
  if (layer.promptScaffold !== undefined) next.promptScaffold = layer.promptScaffold;

  const kind = settings.reasoningCapability?.kind;
  if (kind === "openai" && layer.openaiReasoningEffort !== undefined) {
    next.openaiReasoningEffort = layer.openaiReasoningEffort;
  }
  if (kind === "deepseek") {
    if (layer.deepseekReasoningEffort !== undefined) next.deepseekReasoningEffort = layer.deepseekReasoningEffort;
    if (layer.deepseekThinkingEnabled !== undefined) next.deepseekThinkingEnabled = layer.deepseekThinkingEnabled;
  }
  if (kind === "anthropic-adaptive") {
    if (layer.anthropicThinkingEffort !== undefined) next.anthropicThinkingEffort = layer.anthropicThinkingEffort;
    next.anthropicThinkingEnabled = true;
    next.anthropicThinkingBudget = undefined;
  }
  if (kind === "anthropic-budget") {
    if (layer.anthropicThinkingEnabled !== undefined) next.anthropicThinkingEnabled = layer.anthropicThinkingEnabled;
    if (layer.anthropicThinkingBudget !== undefined) next.anthropicThinkingBudget = layer.anthropicThinkingBudget;
  }
  if (kind === "google" && layer.googleThinkingLevel !== undefined) {
    next.googleThinkingLevel = layer.googleThinkingLevel;
  }

  return next;
}

/// providers.json の役割プロファイルを先に、ユーザーの役割別オーバーライドを後に重ねる(後勝ち)。
export function applyRoleProfile(
  settings: AiSettings,
  profile: ModelRoleProfile | undefined,
  overrides: RoleParamOverrides | undefined,
): AiSettings {
  return applyRoleProfileLayer(applyRoleProfileLayer(settings, profile), overrides);
}

export function resolveChatRunSettings(config: ProviderConfig, settings: AiSettings): AiSettings {
  const resolved = resolveChatSettings(settings);
  const entry = getProviderEntry(config, resolved.provider);
  const usesChatOverride = resolved.provider !== settings.provider || resolved.model !== settings.model;
  return enforceRequiredThinking(
    applyRuntimeModelDefaults(resolved, getProviderModelDefaults(entry, resolved.model), usesChatOverride),
  );
}

export function resolveBackgroundRunSettings(config: ProviderConfig, settings: AiSettings): AiSettings {
  const resolved = resolveBackgroundSettings(settings);
  const entry = getProviderEntry(config, resolved.provider);
  const usesBackgroundOverride =
    resolved.provider !== settings.provider || resolved.model !== settings.model;
  return enforceRequiredThinking(applyRuntimeModelDefaults(
    resolved,
    getProviderModelDefaults(entry, resolved.model),
    usesBackgroundOverride,
  ));
}

export function resolveWritingRunSettings(config: ProviderConfig, settings: AiSettings): AiSettings {
  // 執筆系の選択元を解決してから、モデル既定値と役割上書きを適用する。
  // モデル既定値 → 役割プロファイル → ユーザーの役割別オーバーライド の順で重ねる
  // (トークン上限だけは保存済みのグローバル値を維持する)。
  const resolved = resolveWritingSettings(settings);
  const defaults = getProviderModelDefaults(getProviderEntry(config, resolved.provider), resolved.model);
  const usesOverride = resolved.provider !== settings.provider || resolved.model !== settings.model;
  const base = applyRuntimeModelDefaults(resolved, defaults, usesOverride);
  const withWritingProfile = enforceRequiredThinking(
    applyRoleProfile(base, defaults?.writing, settings.writingOverrides),
  );
  return isThinkingEnabledForRun(withWritingProfile)
    ? withWritingProfile
    : {
        ...withWritingProfile,
        maxTokens: Math.min(withWritingProfile.maxTokens, NON_THINKING_WRITING_MAX_TOKENS),
      };
}

export function resolveJudgmentRunSettings(config: ProviderConfig, settings: AiSettings): AiSettings {
  const resolved = resolveJudgmentSettings(settings);
  const entry = getProviderEntry(config, resolved.provider);
  const defaults = getProviderModelDefaults(entry, resolved.model);
  const usesOverride = resolved.provider !== settings.provider || resolved.model !== settings.model;
  const base = applyRuntimeModelDefaults(resolved, defaults, usesOverride);
  return enforceRequiredThinking(
    applyRoleProfile(base, defaults?.judgment, settings.judgmentOverrides),
  );
}

File: frontend-rs/src/ai/role_settings.rs
//! 実行時のモデル設定解決（チャット/執筆系/判断系/バックグラウンド）。
//! モデル既定値・役割プロファイル・ユーザーオーバーライドを順に重ねる。
//!
//! TypeScript `role-settings.ts` の Rust 移植。

#![allow(dead_code)]
use serde_json::{Map, Value};

use super::capability::{self, ReasoningCapability};

/// DeepSeek V4 モデルの場合、thinking を強制有効化した設定を返す。
pub fn enforce_required_thinking(mut settings: Value) -> Value {
    let model = settings
        .get("model")
        .and_then(Value::as_str)
        .unwrap_or("");
    if capability::is_deepseek_v4_model(model) {
        if let Some(obj) = settings.as_object_mut() {
            obj.insert(
                "deepseekThinkingEnabled".into(),
                Value::Bool(true),
            );
        }
    }
    settings
}

/// モデル既定値を実行時設定に適用する。
///
/// `defaults` はモデルの ProviderModelDefaults（JSON object）。
/// `apply_token_defaults` が true の場合のみ maxTokens/maxContextTokens も上書きする。
pub fn apply_runtime_model_defaults(
    settings: Value,
    defaults: Option<&Value>,
    apply_token_defaults: bool,
) -> Value {
    let Some(defaults) = defaults else {
        return settings;
    };

    let provider = settings
        .get("provider")
        .and_then(Value::as_str)
        .unwrap_or("");
    let model_id = settings
        .get("model")
        .and_then(Value::as_str)
        .unwrap_or("");

    let capability = defaults
        .get("reasoningCapability")
        .and_then(|v| serde_json::from_value::<ReasoningCapability>(v.clone()).ok())
        .or_else(|| capability::get_model_capability(provider, model_id, None));

    let mut next = settings;
    let obj = next.as_object_mut().expect("settings must be an object");

    // サンプリング系パラメータをマージ
    merge_optional_number(obj, defaults, "temperature");
    if apply_token_defaults {
        merge_optional_number(obj, defaults, "maxTokens");
        merge_optional_number(obj, defaults, "maxContextTokens");
    }
    merge_optional_number(obj, defaults, "topP");
    merge_optional_number(obj, defaults, "topK");
    merge_optional_number(obj, defaults, "frequencyPenalty");
    merge_optional_number(obj, defaults, "presencePenalty");

    // reasoningCapability を設定
    if let Some(ref cap) = capability {
        obj.insert(
            "reasoningCapability".into(),
            serde_json::to_value(cap).unwrap_or_default(),
        );
    }

    // プロバイダ種別に応じた reasoning/thinking パラメータを適用
    if let Some(cap) = &capability {
        match cap.kind.as_str() {
            "openai" => {
                merge_optional_string(obj, defaults, "openaiReasoningEffort");
            }
            "deepseek" => {
                merge_optional_string(obj, defaults, "deepseekReasoningEffort");
            }
            "anthropic-adaptive" => {
                obj.insert("anthropicThinkingEnabled".into(), Value::Bool(true));
                obj.insert("anthropicThinkingBudget".into(), Value::Null);
                let default_effort = cap
                    .default_effort
                    .as_deref()
                    .filter(|e| {
                        matches!(*e, "low" | "medium" | "high" | "xhigh" | "max")
                    });
                if let Some(effort) = default_effort {
                    if !obj.contains_key("anthropicThinkingEffort") {
                        obj.insert(
                            "anthropicThinkingEffort".into(),
                            Value::String(effort.into()),
                        );
                    }
                }
                merge_optional_string(obj, defaults, "anthropicThinkingEffort");
            }
            "anthropic-budget" => {
                merge_optional_bool(obj, defaults, "anthropicThinkingEnabled");
                merge_optional_number(obj, defaults, "anthropicThinkingBudget");
                obj.insert("anthropicThinkingEffort".into(), Value::Null);
            }
            "google" => {
                merge_optional_string(obj, defaults, "googleThinkingLevel");
            }
            _ => {}
        }
    }

    next
}

/// providers.json の役割プロファイルとユーザーの役割別オーバーライドを
/// 順に重ねる（後勝ち）。layer が先、overrides が後。
pub fn apply_role_profile(
    mut settings: Value,
    layer: Option<&Value>,
    overrides: Option<&Value>,
) -> Value {
    if let Some(layer) = layer {
        settings = apply_role_profile_layer(settings, layer);
    }
    if let Some(overrides) = overrides {
        settings = apply_role_profile_layer(settings, overrides);
    }
    settings
}

fn apply_role_profile_layer(mut settings: Value, layer: &Value) -> Value {
    let obj = settings.as_object_mut().expect("settings must be an object");
    let layer = layer.as_object();
    let Some(layer) = layer else { return settings };

    // サンプリング系は無条件で上書き
    copy_if_present(obj, layer, "temperature");
    copy_if_present(obj, layer, "topP");
    copy_if_present(obj, layer, "topK");
    copy_if_present(obj, layer, "frequencyPenalty");
    copy_if_present(obj, layer, "presencePenalty");
    copy_if_present(obj, layer, "promptScaffold");

    // reasoning/thinking 系は現在のプロバイダ種別に一致するものだけ適用
    let kind = obj
        .get("reasoningCapability")
        .and_then(|v| v.get("kind"))
        .and_then(Value::as_str)
        .map(|s| s.to_string());

    match kind.as_deref() {
        Some("openai") => {
            copy_if_present(obj, layer, "openaiReasoningEffort");
        }
        Some("deepseek") => {
            copy_if_present(obj, layer, "deepseekReasoningEffort");
            copy_if_present(obj, layer, "deepseekThinkingEnabled");
        }
        Some("anthropic-adaptive") => {
            copy_if_present(obj, layer, "anthropicThinkingEffort");
            obj.insert("anthropicThinkingEnabled".into(), Value::Bool(true));
            obj.insert("anthropicThinkingBudget".into(), Value::Null);
        }
        Some("anthropic-budget") => {
            copy_if_present(obj, layer, "anthropicThinkingEnabled");
            copy_if_present(obj, layer, "anthropicThinkingBudget");
        }
        Some("google") => {
            copy_if_present(obj, layer, "googleThinkingLevel");
        }
        _ => {}
    }

    settings
}

// ---------------------------------------------------------------------------
// 内部ヘルパー
// ---------------------------------------------------------------------------

fn merge_optional_number(obj: &mut Map<String, Value>, defaults: &Value, key: &str) {
    if let Some(v) = defaults.get(key) {
        if !v.is_null() && !obj.contains_key(key) {
            obj.insert(key.into(), v.clone());
        }
    }
}

fn merge_optional_string(obj: &mut Map<String, Value>, defaults: &Value, key: &str) {
    if let Some(v) = defaults.get(key) {
        if v.is_string() && !obj.contains_key(key) {
            obj.insert(key.into(), v.clone());
        }
    }
}

fn merge_optional_bool(obj: &mut Map<String, Value>, defaults: &Value, key: &str) {
    if let Some(v) = defaults.get(key) {
        if v.is_boolean() && !obj.contains_key(key) {
            obj.insert(key.into(), v.clone());
        }
    }
}

fn copy_if_present(
    obj: &mut Map<String, Value>,
    source: &Map<String, Value>,
    key: &str,
) {
    if let Some(v) = source.get(key) {
        obj.insert(key.into(), v.clone());
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn enforces_thinking_for_deepseek_v4() {
        let settings = json!({
            "provider": "deepseek",
            "model": "deepseek-v4-0324",
            "deepseekThinkingEnabled": false
        });
        let result = enforce_required_thinking(settings);
        assert_eq!(
            result["deepseekThinkingEnabled"].as_bool().unwrap(),
            true
        );
    }

    #[test]
    fn does_not_enforce_for_non_v4() {
        let settings = json!({
            "provider": "deepseek",
            "model": "deepseek-v3",
            "deepseekThinkingEnabled": false
        });
        let result = enforce_required_thinking(settings);
        assert_eq!(
            result["deepseekThinkingEnabled"].as_bool().unwrap(),
            false
        );
    }

    #[test]
    fn applies_token_defaults_when_requested() {
        let settings = json!({"provider":"openai","model":"gpt-5.1"});
        let defaults = json!({"maxTokens":4096,"maxContextTokens":128000});
        let result =
            apply_runtime_model_defaults(settings, Some(&defaults), true);
        assert_eq!(result["maxTokens"].as_u64().unwrap(), 4096);
        assert_eq!(result["maxContextTokens"].as_u64().unwrap(), 128000);
    }

    #[test]
    fn skips_token_defaults_when_not_requested() {
        let settings = json!({"provider":"openai","model":"gpt-5.1","maxTokens":2048});
        let defaults = json!({"maxTokens":4096});
        let result =
            apply_runtime_model_defaults(settings, Some(&defaults), false);
        assert_eq!(result["maxTokens"].as_u64().unwrap(), 2048);
    }

    #[test]
    fn role_profile_layer_overrides() {
        let settings = json!({
            "provider":"openai",
            "model":"gpt-5.1",
            "temperature":0.7,
            "reasoningCapability":{"kind":"openai"}
        });
        let layer = json!({"temperature":0.3,"openaiReasoningEffort":"high"});
        let result = apply_role_profile_layer(settings, &layer);
        assert_eq!(result["temperature"].as_f64().unwrap(), 0.3);
        assert_eq!(
            result["openaiReasoningEffort"].as_str().unwrap(),
            "high"
        );
    }
}
